### PR TITLE
[SILGen] Store_borrow into in_guaranteed.

### PR DESCRIPTION
**Pull request by nate-chandler**
Don't create a copy in order to pass a value `@in_guaranteed`.  Just `store_borrow` the value.

rdar://128710064

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please apple silicon benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

File: lib/SILGen/SILGenPoly.cpp [hunk 1]
```diff
@@ -2119,32 +2161,43 @@ class TranslateArguments : public ExpanderBase<TranslateArguments, ParamInfo> {
...
                                      outer, innerTy,
                                      SGFContext(&init));
     if (!innerArg.isInContext()) {
       if (innerArg.isPlusOneOrTrivial(SGF)) {
         innerArg.forwardInto(SGF, Loc, &init);
+      } else if (!innerArg.getType().isAddressOnly(SGF.F) &&
+                 target.isGuaranteed()) {
+        auto &tempInit = target.getGuaranteedInit();
+        auto addr = tempInit.getAddress();
+        SGF.Cleanups.getCleanup(tempInit.getInitializedCleanup())
```
Comment (nate-chandler) [inline lib/SILGen/SILGenPoly.cpp:2193]: Now, to remove some of these hacks.  Maybe via a new Initialization subclass.

**Comment by nate-chandler** [comment]
@swift-ci please apple silicon benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
```
### Performance (arm64): -O

**Regression**               | **OLD**  | **NEW**  | **DELTA** | **RATIO**
:---                         | ---:     | ---:     | ---:   | ---:
RawBuffer.copyContents       | 11.075   | 12.122   | +9.5%  | **0.91x (?)**
&nbsp; | | | |
**Improvement**              | **OLD**  | **NEW**  | **DELTA** | **RATIO**
ArrayAppendGenericStructs    | 1146.667 | 871.429  | -24.0% | **1.32x (?)**
FindString.Loop1.Substring   | 211.0    | 163.923  | -22.3% | **1.29x**
ObjectiveCBridgeToNSArray    | 4396.154 | 3996.429 | -9.1%  | **1.10x (?)**
StringFromLongWholeSubstring | 2.054    | 1.907    | -7.2%  | **1.08x**
DataAppendDataSmallToLarge   | 5696.0   | 5289.362 | -7.1%  | **1.08x (?)**

### Code size: -O


### Performance (arm64): -Osize

**Regression**               | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                         | ---:    | ---:    | ---:    | ---:
StringWithCString2           | 0.0     | 0.002   | +200.0% | **0.33x**
RawBufferCopyBytes           | 11.192  | 12.313  | +10.0%  | **0.91x (?)**
&nbsp; | | | |
**Improvement**              | **OLD** | **NEW** | **DELTA** | **RATIO**
DataCreateMediumArray        | 572.877 | 530.704 | -7.4%   | **1.08x (?)**
StringFromLongWholeSubstring | 1.998   | 1.856   | -7.1%   | **1.08x**

### Code size: -Osize


### Performance (arm64): -Onone

**Regression**                     | **OLD**  | **NEW**  | **DELTA** | **RATIO**
:---                               | ---:     | ---:     | ---:   | ---:
ArrayAppendGenericStructs          | 1000.0   | 1187.273 | +18.7% | **0.84x (?)**
&nbsp; | | | |
**Improvement**                    | **OLD**  | **NEW**  | **DELTA** | **RATIO**
DataAppendDataMediumToLarge        | 6925.0   | 5611.268 | -19.0% | **1.23x (?)**
DataAppendDataLargeToSmall         | 6896.429 | 5904.615 | -14.4% | **1.17x (?)**
DataReplaceLarge                   | 7193.333 | 6382.857 | -11.3% | **1.13x (?)**
NSStringConversion.Rebridge.Medium | 122.474  | 111.286  | -9.1%  | **1.10x (?)**
RawBufferCopyBytes                 | 12.474   | 11.565   | -7.3%  | **1.08x (?)**
DataAppendDataLargeToMedium        | 7066.667 | 6566.667 | -7.1%  | **1.08x (?)**
Set.intersection.Seq.Int100        | 259.75   | 242.0    | -6.8%  | **1.07x (?)**

### Code size: -swiftlibs
```

**Comment by nate-chandler** [comment]
@swift-ci please apple silicon benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
```
### Performance (arm64): -O

**Improvement**              | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                         | ---:    | ---:    | ---:   | ---:
FindString.Loop1.Substring   | 210.909 | 161.231 | -23.6% | **1.31x**
StringFromLongWholeSubstring | 2.053   | 1.857   | -9.5%  | **1.11x**

### Code size: -O


### Performance (arm64): -Osize

**Regression**               | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                         | ---:    | ---:    | ---:   | ---:
ArrayAppendGenericStructs    | 887.333 | 1048.75 | +18.2% | **0.85x (?)**
&nbsp; | | | |
**Improvement**              | **OLD** | **NEW** | **DELTA** | **RATIO**
StringWithCString2           | 0.002   | 0.001   | -33.3% | **1.50x**
ArrayAppendToGeneric         | 179.474 | 162.72  | -9.3%  | **1.10x (?)**
StringFromLongWholeSubstring | 1.998   | 1.855   | -7.2%  | **1.08x**

### Code size: -Osize


### Performance (arm64): -Onone

**Regression**            | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                      | ---:    | ---:    | ---:   | ---:
ArrayAppendToGeneric      | 163.529 | 211.176 | +29.1% | **0.77x (?)**
ArrayAppendGenericStructs | 996.25  | 1119.0  | +12.3% | **0.89x (?)**
&nbsp; | | | |
**Improvement**           | **OLD** | **NEW** | **DELTA** | **RATIO**
RawBufferCopyBytes        | 12.5    | 10.672  | -14.6% | **1.17x (?)**

### Code size: -swiftlibs
```

**Review by jckarter — APPROVED** [review]

**Pull request merged**